plugins/amdgpu/amdgpu_plugin.c

```diff
@@ -910,11 +907,139 @@ static int dump_devices(int fd,
 	return ret;
 }
 
+static int dump_bos(int fd, struct kfd_ioctl_criu_process_info_args *info_args, CriuKfd *e)
+{
+	struct kfd_ioctl_criu_dumper_args args = {0};
+	struct kfd_criu_bo_bucket *bo_buckets;
+	struct thread_data *thread_datas;
+	uint8_t *priv_data;
+	int ret = 0, i;
+
+	pr_debug("Dumping %lld BOs\n", info_args->total_bos);
+
+	thread_datas = xzalloc(sizeof(*thread_datas) * e->num_of_gpus);
+	if (!thread_datas) {
+		ret = -ENOMEM;
+		goto exit;
+	}
+
+	args.type = KFD_CRIU_OBJECT_TYPE_BO;
+	args.objects_index_start = 0; /* Partial object lists not supported for now */
+
+	args.num_objects = info_args->total_bos;
+	args.objects_size = (args.num_objects * sizeof(*bo_buckets)) +
+				info_args->bos_priv_data_size;
+
+	args.objects = (uintptr_t)xmalloc(args.objects_size);
+	if (!args.objects) {
+		ret = -ENOMEM;
+		goto exit;
+	}
+
+	ret = kmtIoctl(fd, AMDKFD_IOC_CRIU_DUMPER, &args);
+	if (ret) {
+		pr_perror("amdgpu_plugin: Failed to call dumper (bos) ioctl");
+		goto exit;
+	}
+
+	bo_buckets = (struct kfd_criu_bo_bucket*)args.objects;
+	/* First private data starts after all buckets */
+	priv_data = (uint8_t *)args.objects + (args.num_objects * sizeof(*bo_buckets));
+
+	e->num_of_bos = info_args->total_bos;
+	ret = allocate_bo_entries(e, args.num_objects, bo_buckets);
```

Comment:
Better to use e->num_of_bos instead of args.num_objects. Currently both are the same. But when we support partial object lists in the future, I think the first call will allocate space for all the objects.

Reply:
ACK